k2/csrc/intersect.cu

```diff
@@ -90,950 +69,840 @@ using namespace intersect_internal;  // NOLINT
-              states_idx1 = state_idx01 - states_idx0x;
-
-      int32_t oshape_idx0x = oshape_row_splits1[states_idx0],
-              oshape_idx01 = oshape_idx0x + t,
-              oshape_idx01x = oshape_row_splits2[oshape_idx01],
-              oshape_idx012 = oshape_idx01x + states_idx1;
-      // note: axis 1 of 'states' corresponds to axis 2 of 'oshape'; it's the
-      // state index.  Also,
-
-      keep_states_data[oshape_idx012] = keep_this_state;
-      if (!keep_this_state) {
-        // The reason we set the backward_loglike to -infinity here if it's
-        // outside the beam, is to prevent disconnected states from appearing
-        // after pruning due to numerical roundoff effects near the boundary
-        // at `-beam`.  It would otherwise be correct and harmless to omit
-        // this if-block.
-        backward_loglike = minus_inf;
-      }
-      info->backward_loglike = backward_loglike;
+    float *backward_state_scores_t0 = steps_[0].backward_state_scores.Data();
+    FsaInfo *fsa_info_data = fsa_info_.Data();
+    Array1<float> score_cutoffs(c_, num_fsas_);
+    float *score_cutoffs_data = score_cutoffs.Data();
+    float output_beam = output_beam_;
+    auto lambda_set_cutoffs = [=] __host__ __device__ (int32_t fsa_idx0) -> void {
+      FsaInfo fsa_info = fsa_info_data[fsa_idx0];
+      float tot_score_start = backward_state_scores_t0[0],
+        tot_score_end = forward_state_scores_data[fsa_info.T]
+                 [fsa_info.state_offset + fsa_info.num_states - 1],
+        tot_score_avg = 0.5 * (tot_score_start + tot_score_end),
+        tot_score_diff = fabs(tot_score_end - tot_score_start);
+      // TODO(dan): remove the following after the code is tested.
+      K2_CHECK(tot_score_diff < 0.1);
+      // subtracting the difference in scores is to help make sure we don't completely prune
+      // away all states.
+      score_cutoffs_data[fsa_idx0] = tot_score_avg - tot_score_diff - output_beam;
     };
-    Eval(c_, cur_frame->states.NumElements(), lambda_set_state_backward_prob);
+    Eval(c_, num_fsas_, lambda_set_cutoffs);
+    K2_LOG(INFO) << "Cutoffs = " << score_cutoffs;
+    return score_cutoffs;
   }
 
+
+
   ContextPtr c_;
-  FsaVec &a_fsas_;         // Note: a_fsas_ has 3 axes.
-  int32_t a_fsas_stride_;  // 1 if we use a different FSA per sequence
-                           // (a_fsas_.Dim0() > 1), 0 if the decoding graph is
-                           // shared (a_fsas_.Dim0() == 1).
+  FsaVec &a_fsas_;          // a_fsas_: decoding graphs, with same Dim0() as
+                            // b_fsas_. Note: a_fsas_ has 3 axes.
+
   DenseFsaVec &b_fsas_;
-  float search_beam_;
+
+  // num_fsas_ equals b_fsas_.shape.Dim0() == a_fsas_.Dim0().
+  int32_t num_fsas_;
+
+  // This is just a copy of a_fsas_.arcs, with a couple extra pieces of information.
+  struct CompressedArc {
+    // src_state of Arc, as uint16 (an idx1)
+    uint16_t src_state;
```

Comment:
Will `uint16_t` limit the size of the input FsaVec？
It cannot contain more than (1<<16) == 64 k states.

Reply:
It does limit it.  
This version of the code is not suitable for intersection with large FSAs though, because there is no pruning.
So in practice it's not much loss.